mini-alloc/src/imp.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright 2023, Offchain Labs, Inc.
+// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/stylus/licenses/COPYRIGHT.md
+
+use core::{
+    alloc::{GlobalAlloc, Layout},
+    arch::wasm32,
+    num::NonZeroUsize as NonZero,
+};
+
+pub struct MiniAlloc;
+
+/// This is not a valid implementation of [`Sync`] but is ok in single-threaded WASM.
+unsafe impl Sync for MiniAlloc {}
+
+impl MiniAlloc {
+    pub const INIT: Self = MiniAlloc;
+}
+
+unsafe impl GlobalAlloc for MiniAlloc {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        alloc_impl(layout).unwrap_or(core::ptr::null_mut())
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+}
+
+extern "C" {
+    /// This symbol is created by the LLVM linker.
+    static __heap_base: u8;
+}
+
+/// The WASM page size, or 2^16 bytes.
+const PAGE_SIZE: usize = 1 << 16;
+
+/// Represents the negation of the allocator's bump offset and boundary
+///
+/// We store the negation because we can align the negative offset in fewer
+/// instructions than the positive offset.
+static mut STATE: Option<(NonZero, usize)> = None;
+
+unsafe fn alloc_impl(layout: Layout) -> Option<*mut u8> {
```

Comment:
Hi @mikebenfield should we validate `layout.size()` before doing more stuff here? Maybe check it is non-zero or not some excessive value